beacon_node/network/src/sync/sampling.rs

```diff
@@ -0,0 +1,486 @@
+        self.requests.values().map(|r| r.block_root).collect()
+    }
+
+    pub fn on_new_sample_request(
+        &mut self,
+        block_root: Hash256,
+        block_slot: Slot,
+        cx: &mut SyncNetworkContext<T>,
+    ) -> Option<(SamplingRequester, SamplingResult)> {
+        let requester = SamplingRequester::ImportedBlock(block_root);
+
+        let request = match self.requests.entry(requester) {
+            Entry::Vacant(e) => e.insert(ActiveSamplingRequest::new(
+                block_root,
+                block_slot,
+                requester,
+                &self.sampling_config,
+                self.log.clone(),
+            )),
+            Entry::Occupied(_) => {
+                warn!(self.log, "Ignoring duplicate sampling request"; "id" => ?requester);
+                return None;
+            }
+        };
+
+        debug!(self.log, "Created new sample request"; "id" => ?requester);
+
+        // TOOD(das): If a node has very little peers, continue_sampling() will attempt to find enough
+        // to sample here, immediately failing the sampling request. There should be some grace
+        // period to allow the peer manager to find custody peers.
+        request
+            .continue_sampling(cx)
+            .transpose()
+            .map(|result| (requester, result))
+    }
+
+    pub fn on_sample_downloaded(
+        &mut self,
+        id: SamplingId,
+        peer_id: PeerId,
+        resp: Result<(DataColumnSidecarList<T::EthSpec>, Duration), LookupFailure>,
+        cx: &mut SyncNetworkContext<T>,
+    ) -> Option<SamplingResult> {
+        let Some(request) = self.requests.get_mut(&id.id) else {
+            // TOOD(das): This log can happen if the request is error'ed early and dropped
+            debug!(self.log, "Sample downloaded event for unknown request"; "id" => ?id);
+            return None;
+        };
+
+        let result = request.on_sample_downloaded(peer_id, id.column_index, resp, cx);
+        self.handle_sampling_result(result, &id.id)
+    }
+
+    pub fn on_sample_verified(
+        &mut self,
+        id: SamplingId,
+        result: Result<(), String>,
+        cx: &mut SyncNetworkContext<T>,
+    ) -> Option<SamplingResult> {
+        let Some(request) = self.requests.get_mut(&id.id) else {
+            // TOOD(das): This log can happen if the request is error'ed early and dropped
+            debug!(self.log, "Sample verified event for unknown request"; "id" => ?id);
+            return None;
+        };
+
+        let result = request.on_sample_verified(id.column_index, result, cx);
+        self.handle_sampling_result(result, &id.id)
+    }
+
+    fn handle_sampling_result(
+        &mut self,
+        result: Result<Option<()>, SamplingError>,
+        id: &SamplingRequester,
+    ) -> Option<SamplingResult> {
+        let result = result.transpose();
+        if result.is_some() {
+            debug!(self.log, "Removed sampling request"; "id" => ?id);
```

Comment:
Maybe some reasons on why its removed would be useful, e.g. "Remove completed sampling request" (completed means success/failed here)